### PR TITLE
[FIX] ComponentAnalyzer: Detect model types from dataSource

### PR DESCRIPTION
**Pull request by matz3**

### CODE REVIEW

File: lib/lbt/analyzer/ComponentAnalyzer.js [hunk 1]
```diff
@@ -116,12 +117,43 @@ class ComponentAnalyzer {
...
+				if (!oDataSource) {
+					return;
+				}
+				// default dataSource type is OData
+				const dataSourceType = oDataSource.type || "OData";
+				switch (dataSourceType) {
+				case "OData":
+					if (oDataSource.settings && oDataSource.settings.odataVersion === "4.0") {
+						modelType = "sap.ui.model.odata.v4.ODataModel";
+					} else {
```
Comment (codeworrior) [inline lib/lbt/analyzer/ComponentAnalyzer.js]: not yet relevant, but should be limited to odataVersion === "2.0" or undefined.

File: test/lib/lbt/analyzer/ComponentAnalyzer.js [hunk 2]
```diff
@@ -594,3 +594,270 @@ test("_analyzeManifest: Manifest with models", async (t) => {
...
+
+	const analyzer = new ComponentAnalyzer();
+	await analyzer._analyzeManifest(manifest, moduleInfo);
+
+	t.true(stubAddDependency.calledOnce, "addDependency was called once");
+	t.deepEqual(stubAddDependency.getCall(0).args[0], "sap/ui/model/json/JSONModel.js",
+		"addDependency should be called with the dependency name");
+});
+
+test("_analyzeManifest: Manifest with V4 OData model via dataSources", async (t) => {
```
Comment (RandomByte) [inline test/lib/lbt/analyzer/ComponentAnalyzer.js]: "V4 OData" should be "XML"

File: lib/lbt/analyzer/ComponentAnalyzer.js (hunk 1 above)
Comment (matz3) [reply lib/lbt/analyzer/ComponentAnalyzer.js]: Good point. I'll fix it here and in the runtime.

File: test/lib/lbt/analyzer/ComponentAnalyzer.js (hunk 2 above)
Comment (matz3) [reply test/lib/lbt/analyzer/ComponentAnalyzer.js]: Done

File: lib/lbt/analyzer/ComponentAnalyzer.js [hunk 3]
```diff
@@ -116,12 +117,47 @@ class ComponentAnalyzer {
...
+			let modelType;
 			if ( options.type ) {
-				const module = ModuleName.fromUI5LegacyName( options.type );
-				log.verbose("derived model implementation dependency ", module);
-				info.addDependency(module);
+				modelType = options.type;
+			} else if ( options.dataSource && sapApp.dataSources ) {
+				const oDataSource = sapApp.dataSources[options.dataSource];
+				if (!oDataSource) {
+					return;
```
Comment (tommyvinhlam) [inline lib/lbt/analyzer/ComponentAnalyzer.js]: Maybe log a warning that a model is declared in the manifest which can't be matched?

Comment (tommyvinhlam) [reply lib/lbt/analyzer/ComponentAnalyzer.js]:
```suggestion
log.warning(`Provided dataSource ${options.dataSource} for model "${model}" is unknown.`);
return;
```

File: lib/lbt/analyzer/ComponentAnalyzer.js [hunk 4]
```diff
@@ -116,12 +117,47 @@ class ComponentAnalyzer {
...
+				const dataSourceType = oDataSource.type || "OData";
+				switch (dataSourceType) {
+				case "OData":
+					if (oDataSource.settings && oDataSource.settings.odataVersion === "4.0") {
+						modelType = "sap.ui.model.odata.v4.ODataModel";
+					} else if (!oDataSource.settings || oDataSource.settings.odataVersion === "2.0") {
+						// Default if no version is specified
+						modelType = "sap.ui.model.odata.v2.ODataModel";
+					} else {
+						// Unknown version
```
Comment (tommyvinhlam) [inline lib/lbt/analyzer/ComponentAnalyzer.js]:
```suggestion
						log.warning(`Provided OData version in dataSource "${oDataSource.settings.odataVersion}" is unknown.`);
```

File: lib/lbt/analyzer/ComponentAnalyzer.js [hunk 5]
```diff
@@ -116,12 +117,47 @@ class ComponentAnalyzer {
...
+				if (!oDataSource) {
+					return;
+				}
+				// default dataSource type is OData
+				const dataSourceType = oDataSource.type || "OData";
+				switch (dataSourceType) {
+				case "OData":
+					if (oDataSource.settings && oDataSource.settings.odataVersion === "4.0") {
+						modelType = "sap.ui.model.odata.v4.ODataModel";
+					} else if (!oDataSource.settings || oDataSource.settings.odataVersion === "2.0") {
```
Comment (tommyvinhlam) [inline lib/lbt/analyzer/ComponentAnalyzer.js]:
```suggestion
					} else if (!oDataSource.settings || !oDataSource.settings.odataVersion || oDataSource.settings.odataVersion === "2.0") {
```

File: lib/lbt/analyzer/ComponentAnalyzer.js [hunk 6]
```diff
@@ -116,12 +117,47 @@ class ComponentAnalyzer {
...
+					}
+					break;
+				case "JSON":
+					modelType = "sap.ui.model.json.JSONModel";
+					break;
+				case "XML":
+					modelType = "sap.ui.model.xml.XMLModel";
+					break;
+				default:
+					// for custom dataSource types, the class should already be specified in the sap.ui5 models config
```
Comment (tommyvinhlam) [inline lib/lbt/analyzer/ComponentAnalyzer.js]: log.warn(`Unknown dataSource type "${dataSourceType}".`)

File: lib/lbt/analyzer/ComponentAnalyzer.js [hunk 7]
```diff
@@ -116,12 +117,47 @@ class ComponentAnalyzer {
...
+					return;
+				}
+			} else {
+				// no type and no dataSource defined
```
Comment (tommyvinhlam) [inline lib/lbt/analyzer/ComponentAnalyzer.js]:
```suggestion
				log.warn(`No model type and no dataSource defined.`)
```

**Pull request merged**